Add proposal for node maintenance.

mml opened the pull request:
AKA draining

mml committed:
Add proposal for node maintenance.

AKA draining

mml commented:
This still needs some editing, but putting it out there sooner than later.

mml commented:
@davidopp pointed out that I shouldn't be using events for anything important, so that will be revised shortly.

davidopp commented:
@bgrant0607 this was auto-assigned to you but I'm happy to do the first review if you're busy

ncdc commented:
cc @kubernetes/rh-cluster-infra @smarterclayton @pravisankar @timothysc

pmorie commented on docs/proposals/maintenance.md:
```diff
@@ -0,0 +1,179 @@
+
+## Service Quality
+
+This may be thought of as the cell admin offering a guarantee or SLO, or as the
```
s/cell/node, or am i missing something?

mml replied:
On Wed, Nov 18, 2015 at 10:18 PM, Paul Morie <notifications@github.com>
wrote:

> In docs/proposals/maintenance.md
> <https://github.com/kubernetes/kubernetes/pull/17393#discussion_r45305996>
> :
>
> > +
> > +Nodes need periodic maintenance, most of which require partial or total
> > +unavailability to pods.  The cluster admin would like for this to happen (1)
> > +automatically, and (2) respecting any service quality constraints.
> > +
> > +Examples of maintenance:
> > +
> > +- unmount/remount filesystems to change mkfs params
> > +- remove from service but leave up for debugging
> > +- reboot for kernel or image upgrade
> > +- remove from service for hardware repair
> > +- remove from service permanently
> > +
> > +## Service Quality
> > +
> > +This may be thought of as the cell admin offering a guarantee or SLO, or as the
>
> s/cell/node, or am i missing something?
>

Fixed to 'cluster admin'.  Borg terminology was seeping in.


> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/17393/files#r45305996>.
>

timothysc replied:
@pmorie xref to google nomenclature: http://static.googleusercontent.com/media/research.google.com/en//pubs/archive/43438.pdf

pmorie commented on docs/proposals/maintenance.md:
````diff
@@ -0,0 +1,179 @@
+
+```
+type DrainSpec struct {
+  LameNodes    PodList
````
Suggest trying to find a different name than `lame`, which might have connotations that are undesirable.  How about referring to this state as `reduced-availability`?

mml replied:
On Wed, Nov 18, 2015 at 10:24 PM, Paul Morie <notifications@github.com>
wrote:

> In docs/proposals/maintenance.md
> <https://github.com/kubernetes/kubernetes/pull/17393#discussion_r45306233>
> :
>
> > +
> > +### Implementation of the above as a controller
> > +
> > +The parameters of a specific lame, drain, or unlame request could be pushed to
> > +the API server as a drain object.  The drain controller would have to manage
> > +multiple drains in flight at once.  If the drains have differing parameters,
> > +then more aggressive drains would generally starve safer ones.  While this could
> > +be a useful feature in emergencies, it will be a lot easier to reason about
> > +multiple drains in progress if they all share the same parameters, which would
> > +be configured via kube-env and would represent the cell admin's SLO.
> > +
> > +The drain object would thus look something like this
> > +
> > +```
> > +type DrainSpec struct {
> > +  LameNodes    PodList
>
> Suggest trying to find a different name than lame, which might have
> connotations that are undesirable. How about referring to this state as
> reduced-availability?
>

I would like a term that cleanly has two verbs and a noun describing the
state.  Do you have any suggestions along those lines?  In the current
case: lame, unlame, and lamed.

Some suggestions from discussions with @bgrant0607 and @erictune

- cordon, uncordon, cordoned
- decommission, recommission, decommissioned
- close, open, closed (as in a faucet or business)
- shut, reopen, shut (ditto)

—
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/17393/files#r45306233>.
>

swagiaal replied:
> - decommission, recommission, decommissioned

+1

mml replied:
> decommission, recommission, decommissioned
> +1

The problem with most of the examples we came up with (cordon excepted) is that they sound an awful lot like what `drain` actually does, which is truly remove the node from service, not just mark it unschedulable.

pmorie replied:
I really like cordon and uncordon.  I'm going to move this comment to be on the PR instead of a diff so it's easily findable.

pmorie commented on docs/proposals/maintenance.md:
```diff
@@ -0,0 +1,179 @@
+
+### Varying SLOs
+
+We could implement multiple SLOs, either by allowing the cell admin to specify
```
It's unclear what cell means here.

timothysc replied:
s/cell/cluster

timothysc commented on docs/proposals/maintenance.md:
```diff
@@ -0,0 +1,179 @@
+- reboot for kernel or image upgrade
+- remove from service for hardware repair
+- remove from service permanently
+
```
Please highlight specific use cases around this maintenance primitive.  A good example is taking down a group of nodes, e.g. a rack for maintenance.

mml replied:
> Please highlight specific use cases around this maintenance primitive. A good example is taking down a group of nodes, e.g. a rack for maintenance.

Not sure I understand.  What's missing from the examples I gave?  Are you looking for examples that would benefit specifically from the ability to specify a *group* of nodes at once?

> Or this could also potentially be used indirectly to coalesce a cloud deployment after a burst.

To clarify:  This means the user brought up a bunch of VMs and now they want to bring them down, right?

timothysc replied:
> Are you looking for examples that would benefit specifically from the ability to specify a group of nodes at once?

Yes, as it outlines the motivation from the user/admin story. 

> To clarify: This means the user brought up a bunch of VMs and now they want to bring them down, right?

Yup

mml replied:
Added.

timothysc commented:
I see no notion of forgiveness/claim-life here, could we at least reference such that once available we can address.

timothysc commented on docs/proposals/maintenance.md:
```diff
@@ -0,0 +1,179 @@
+### Reliable Event Log
+
+If we wish our SLO to cover durations longer than an hour or so, the loss of
+historical state on restart becomes a real problem.  Adding a reliable event log
```
s/event log/history service

mml replied:
> s/event log/history service

Done.

mml commented:
On Thu, Nov 19, 2015 at 1:13 PM, Timothy St. Clair <notifications@github.com
> wrote:

> I see no notion of forgiveness/claim-life here, could we at least
> reference such that once available we can address.
>

I understand what "forgiveness " means, but not "claim-life".  Can you more
specifically say what you're driving at?

timothysc commented:
@mml there are no points on the proposal that outline how pods which may been matched based on predicate will maintain their place after maintenance.   Namely we are going to want data persistance across maintenance windows in the future. 

/cc @kubernetes/rh-storage

mml commented:
On Thu, Nov 19, 2015 at 2:17 PM, Timothy St. Clair <notifications@github.com
> wrote:

> @mml <https://github.com/mml> there are no points on the proposal that
> outline how pods which may been matched based on predicate will maintain
> their place after maintenance. Namely we are going to want data persistance
> across maintenance windows in the future.
>

I think I understand.  An example would be storage servers that need to go
back onto the machine that has the specific bytes they serve up.

How will such pods managed in general?  Intentional removal for maintenance
is just one reason the pod might die.

Also, is there an issue covering this use case?

/cc @kubernetes/rh-storage
> <https://github.com/orgs/kubernetes/teams/rh-storage>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/17393#issuecomment-158217978>
> .
>

timothysc commented:
https://github.com/kubernetes/kubernetes/issues/1574

mml commented:
On Thu, Nov 19, 2015 at 2:39 PM, Timothy St. Clair <notifications@github.com
> wrote:

> #1574 <https://github.com/kubernetes/kubernetes/issues/1574>
>

OK.  It sounds like forgiveness, if implemented, would cover one case:
brief removal from service for a reboot.  You would simply configure
storage pods to forgive these events.

But, quoting @bgrant0607

> This approach would generalize nicely for cases where, for example,
applications wanted to endure reboots but give up in the case of extended
outages or in the case that the disk goes bad.

If you're taking a node down forever, you may have to get unreplicated
bytes off.  *If* we need to distinguish this type of event and handle it
differently (this may not be a use case anyone has yet), we should file an
issue describing it.

swagiaal commented on docs/proposals/maintenance.md:
```diff
@@ -0,0 +1,179 @@
+parameters:
+
+- time between evictions (of pods in the same replica set)
+- minimum shard strength (minimum fraction of pods healthy)
```
Hmm.. what does "fraction of pods healthy" mean ?

mml replied:
On Fri, Nov 20, 2015 at 6:52 AM, Sami Wagiaalla <notifications@github.com>
wrote:

> Hmm.. what does "fraction of pods healthy" mean ?
>
What fraction of pods in the ReplicaSet are scheduled and have Ready=True,
assuming they have readiness probes defined.  Would "ready" be a better
term than "healthy" here?

davidopp replied:
you should probably add "of pods in the same replica set" as you did in the previous line

thockin replied:
The use of "shard" is a little confusing to me.  We don't have a mapping of shards to ReplicationControllers, much less multi-dimensionally (e.g. 10 shards, 3 copies of each).  Calling it "shard strength" makes me think of the second dimension (e.g. it's ok to have only 1 copy of my shard as long as you don't do that to more than 1 shard at a time) but we have (AFAIK) no such concept.

mml replied:
> The use of "shard" is a little confusing to me.

Agreed.  How about "replica strength"?  Considered "pod strength", but that's weaker.

swagiaal commented on docs/proposals/maintenance.md:
```diff
@@ -0,0 +1,179 @@
+
+- time between evictions (of pods in the same replica set)
+- minimum shard strength (minimum fraction of pods healthy)
+- eviction notice
```
Is this the time between the notice and forced eviction or whether notice should be given ?

davidopp replied:
@swagiaal Yes, this is already part of the API, see the TerminationGracePeriodSeconds in PodSpec

swagiaal commented:
> An example would be storage servers that need to go back onto the machine that has the specific bytes they serve up. How will such pods managed in general? Intentional removal for maintenance is just one reason the pod might die.

I think that depends on the strength of the nodeSelector used to schedule the pod in the first place. If the node selector can only be fulfilled by the node which has been decommissioned then the pod will wait until that node is back. If not the pod will be scheduled to a new node and local storage will be lost. I think that is fair enough because the nodeSelector reflects the user's desire. So an application which uses node storage as computation cache can be safely rescheduled to a new node but will be a bit slow at first, and an application which uses node storage for important state will not be rescheduled until the node is back up. User's can use nodeSelectors to express that. Network storage will of course follow the pod to the new node.

erictune commented on docs/proposals/maintenance.md:
```diff
@@ -0,0 +1,179 @@
+- time between evictions (of pods in the same replica set)
+- minimum shard strength (minimum fraction of pods healthy)
+- eviction notice
+
```
Is there an implicit parameter here, which is "maximum machines drained at once", which is equal to the number of concurrent drains.  Maybe I want to drain N machines at once in a large cluster?  Or maybe it is better expressed as "min machines still in cluster at any time"?  Gets complicated if we add dedicated machines though, so maybe too complex? should be explicit (allow draining two at once

davidopp replied:
Why would the user ever care about that, as long as the system is satisfying the shard strength and eviction rate requirements of all the pods on the machines you are simultaneously draining? Is it to leave slack capacity for workload bursts? Seems like that should be a cluster autoscaling parameter rather than something here.

mml replied:
On Fri, Nov 20, 2015 at 7:47 AM, Eric Tune <notifications@github.com> wrote:

> Is there an implicit parameter here, which is "maximum machines drained at
> once", which is equal to the number of concurrent drains. Maybe I want to
> drain N machines at once in a large cluster? Or maybe it is better
> expressed as "min machines still in cluster at any time"? Gets complicated
> if we add dedicated machines though, so maybe too complex? should be
> explicit (allow draining two at once
>
We sort of implicitly manage it by guaranteeing that only a limited amount
of work is pending, but we don't offer any guarantees on slack capacity.
To give a concrete example of why this matters: you could imagine someone
downing a replication controller for an (unmanaged) upgrade, then the drain
controller sees the green light and is able to shut down a bunch of nodes,
then the user ups a new rc and now there is way less available capacity and
things go pending.

You could manage this with max-nodes-in-flight, as you suggest, or with a
"percentage of capacity" parameter.  You'd also get defense in depth from
rate limiting pod deletion rates.  If the controller goes slow, it can only
do so much damage during the window.

Either way, I will add a mention of this issue.

smarterclayton replied:
I expect this to be requested frequently by admins who have strict policies on cluster management.

erictune replied:
maximum machines drained at once is an issue for bare metal.  Suppose an HPA scales up an RC due to increased load.   There is no such thing as cluster autoscaling on bare metal.

erictune commented on docs/proposals/maintenance.md:
````diff
@@ -0,0 +1,179 @@
+  --min-shard-strength=0.75 \
+  --min-seconds-between-evictions=900 \
+  --grace=900 $nodes
+```
````
Can you provide an example of how a user would use this in conjunction with the actual maintenance script?

For example, if I want to run the command `gcloud instances ssh $HOST -- "apt-get update && shutdown -r now"`, how does that integrate with the drain command?  Does the drain command run a script for me after draining each machine?  Or do I have to watch all the nodes and guess which one was drained by the kubectl drain command?  What if there are multiple drains, how do I know which one was intended for me to operate on, versus some other process?

davidopp replied:
I think having the drain command run a script is bundling too much into a single operation (and too speculative as to what users really want to have happen after draining). But you bring up a good point. Perhaps the machine should start reporting some new status once it is drained. Is this an appropriate usage of the NodePhase "Terminated"? The comment there says "NodeTerminated means the node has been removed from the cluster" so maybe not. (Though AFAICT we're not actually using NodeTerminated currently, so we could just change the comment and start using it this way...) Alternatively we could have a new NodeCondition called NodeDrained or something like that. But in either approach, how does a node know it is drained, and how does it know when it is undrained? A third approach would be to put a bool into the NodeSpec that `kubectl drain` would set and `kubectl undrain` would un-set. It seems less than ideal but it's the best option I can think of at the moment. It would also address your "multiple drains" question if it were a string (or something like that) instead of a bool.

mml replied:
This is a good point.  The simple command line version makes building a simple drain/{reboot or whatever}/unlame workflow easy, but you totally lose that with all the more complicated versions I described.

I'm sure we can figure out a set of CLI arguments that would DTRT, but the problem gets even thornier when it's supposed to run unattended on a controller.  I'll think about this some more.

erictune replied:
I was assuming that the "drain multiple" case meant "drain one at a time and then return to service, and then go on to the next one".  But, maybe you didn't mean that.  Draining machines in blocks is useful if you have a lot of machines or need to work on several at once (e.g. drain a rack).  So, I think "drain multiple in parallel" is a pretty useful function.  The "drain multiple in series" is less obviously useful since I don't see yet how you integrate that with the other steps of the maintenance.

erictune commented on docs/proposals/maintenance.md:
```diff
@@ -0,0 +1,179 @@
+pods in replication sets also impacted by the drain, it can hold off on
+evictions impacting that replication set until MinSecondsBetweenEvictions and
+MinShardStrength are OK again.
+
```
Same question as above: how do I synchronize my actual maintenance commands (install, reboot, unmount, etc) with the draining cycle.

erictune commented on docs/proposals/maintenance.md:
```diff
@@ -0,0 +1,179 @@
+type DrainSpec struct {
+  LameNodes    PodList
+  RemoveNodes  PodList
+  AddNodes     PodList
```
Would it be better to encourage one DrainSpec per type of maintenance?  If so, then maybe

```
type DrainSpec struct {
Nodes NodeList
Action string // one of "Lame", "Remove", "Add"
}
```

davidopp replied:
+1 to @erictune 's suggestion
I would suggest "drain" and "undrain" instead of remove/add though, to make it clear the node is still "in the cluster"

davidopp replied:
Also, what happened to shard-strength and time-between-evictions? You had it in the spec for the command-line version but not in this object.

mml replied:
On Fri, Nov 20, 2015 at 3:28 PM, David Oppenheimer <notifications@github.com
> wrote:

> Also, what happened to shard-strength and time-between-evictions? You had
> it in the spec for the command-line version but not in this object.
>

Search for "If the drains have differing parameters"

mml replied:
> Would it be better to encourage one DrainSpec per type of maintenance

Absolutely.  Done.  Way easier to reason about and explain.

erictune commented on docs/proposals/maintenance.md:
```diff
@@ -0,0 +1,179 @@
+The implementation is the same whether there is one drain in flight or several.
+The drain agent (either kubectl or the drain controller) must watch for
+replication controllers and pods, and it will watch all pods to notice any time
+there is a disruption to a ReplicaSet.  Because there is no reliable mechanism
```
I think you can do this without watching for disruptions.  It can just watch the current state: watch all RCs, check that each RC has as at least `drain.minShardStrength * rc.spec.replicas` pods, all of which are at least `minSecondsBetweenEvictions` old.

erictune commented on docs/proposals/maintenance.md:
```diff
@@ -0,0 +1,179 @@
+
+Rather than only waiting `X` seconds between evictions, an admin might want to
+limit the total in, say, the trailing 6 or 24 hour window.  This requires more
+reliable event logging, described below.
```
This can probably be inferred by taking all the pods in the system, sorting by age, and looking at how many fall into your window.

davidopp replied:
@erictune I don't understand. Pods may have been deleted so I think there's not enough information.

erictune commented:
Good introduction to the problem, staging of work, and parsimony with parameters.  Like it.  A couple of questions and a few optional suggestions offered.

davidopp commented on docs/proposals/maintenance.md:
````diff
@@ -0,0 +1,179 @@
+  --grace=900 $nodes
+```
+
+This would still be pretty simple.  We offer no guarantees about safety in the
````
You should probably say something about how you would handle resuming the drain if the client is accidentally terminated and is then restarted. Or at least what would happen if someone does control-C in the middle of a `kubectl drain` and then runs `kubectl drain` again, even if you don't want to gracefully handle the scenario.

mml replied:
> what would happen if someone does control-C in the middle of a kubectl drain and then runs kubectl drain again

Added an explanation.

davidopp commented on docs/proposals/maintenance.md:
```diff
@@ -0,0 +1,179 @@
+be drained according to the given parameters, and nodes in `AddNodes` would be
+undrained.  The object would be rejected if the sets overlap at all.  The action
+of the drain controller can be paused or stopped by deleting the drain object,
+and restarted by re-posting it.
```
restarting seems to require some way of knowing which machines are already in drained state (we already know which are in lamed state, namely the unschedulable flag, though I'm a little concerned we're overloading it), which gets back to the earlier issue of does a machine signal it is in drained state via status, or phase, or spec...

mml replied:
The machine's drained when it has no pods left.

davidopp commented on docs/proposals/maintenance.md:
```diff
@@ -0,0 +1,179 @@
+of the drain controller can be paused or stopped by deleting the drain object,
+and restarted by re-posting it.
+
+Undo would be trickier, but a clever client could be written such that it
```
do we need an explicit undo? could you just delete the existing DrainSpec and replace it with one where you populate the "undrain" nodes with the set that was previously in "drain" and if the controller hadn't actually gotten around to draining a node yet then the undrain is a no-op

same with undo lame (actually, why don't you have unlame?)

mml replied:
> do we need an explicit undo? could you just delete the existing DrainSpec and replace it with one where you populate the "undrain" nodes with the set that was previously in "drain" and if the controller hadn't actually gotten around to draining a node yet then the undrain is a no-op

Not sure why I mentioned undo.  I think it's better left unmentioned, actually.

> same with undo lame (actually, why don't you have unlame?)

That's what AddNodes is.  I'll tighten up the names there so that it's {Lame,Drain,Unlame}Nodes.

davidopp commented on docs/proposals/maintenance.md:
```diff
@@ -0,0 +1,179 @@
+
+### Varying SLOs
+
+We could implement multiple SLOs, either by allowing the cluster admin to specify
```
What you describe in this paragraph is reasonable but I'd suggest mentioning it briefly also in the Service Quality section -- say that time between eviction and shard strength are cluster-wide properties not per-collection. You could also say that eviction notice is optional -- it defaults to the PodSpec.TerminationGracePeriodSeconds if it is not specified (either in the kubectl version or the object-on-the-server version).

bgrant0607 commented on docs/proposals/maintenance.md:
```diff
@@ -0,0 +1,179 @@
+the similar parameters already present in Deployment.  I believe we should offer
+this safety to pods regardless of whether they're managed by Deployment or not,
+so I don't think Deployment is the right place to pull this from.  Rather, when
+a sketch for multiple or parameterized SLOs as a feature is present, we should
```
http://issues.k8s.io/12611

mml commented:
Pushed a new revision.

davidopp commented on docs/proposals/maintenance.md:
```diff
@@ -0,0 +1,216 @@
+- reboot for kernel or image upgrade
+- remove from service for hardware repair
+- gracefully reducing the size of the cluster
+- take down machines all machines in a rack for maintenance
```
s/machines all machines/all machines/

smarterclayton replied:
Update Docker (or other container managers that can't restart gracefully) without reboot

davidopp commented on docs/proposals/maintenance.md:
```diff
@@ -0,0 +1,216 @@
+
+This is implemented in https://github.com/kubernetes/kubernetes/pull/16698
+
+### One-shot drain of multiple nodes
```
I would title this 'Client-driven drain of multiple nodes' instead
You could use the 'one-shot' term in the last sentence, which I think it what you're referring to (that if you interrupt and resume, it doesn't work quite correctly, so you're supposed to use it as a one-shot).

davidopp commented on docs/proposals/maintenance.md:
```diff
@@ -0,0 +1,216 @@
+kubectl drain \
+  --min-shard-strength=0.75 \
+  --min-seconds-between-evictions=900 \
+  --grace=900 $nodes
```
Is --grace mandatory, or is it optional and when omitted PodSpec. TerminationGracePeriodSeconds is used? In any event, maybe explain. (Actually, maybe all arguments are optional -- I assume that when min-shard-strength and/or min-seconds-between-evictions are omitted, they default to 0?)

smarterclayton replied:
Grace is optional, and should probably not "just work" without taking into account the actual grace period of the pods themselves.  If I --grace-period=900 a node that has a pod with grace period 100000, an admin should not be able to blindly force that pod to 900.  we should required a --confirm in those cases

mml replied:
The usual reason an admin would specify a specific grace period is because they have a near-immediate need for the machines and they aren't willing to wait longer than X seconds for them.  Thus it should be optional, but it could be renamed to --grace-upper-bound and the semantics are that we wait as long as the pod asked for, up to --grace-upper-bound seconds.

davidopp commented on docs/proposals/maintenance.md:
```diff
@@ -0,0 +1,216 @@
+
+### Implementation of the above as a controller
+
+The parameters of a specific lame, drain, or unlame request could be pushed to
```
This is the first time you use these terms in the doc, so you should define them here.

davidopp commented on docs/proposals/maintenance.md:
```diff
@@ -0,0 +1,216 @@
+
+The parameters of a specific lame, drain, or unlame request could be pushed to
+the API server as a drain object.  The drain controller would have to manage
+multiple drains in flight at once.  If the drains have differing parameters,
```
I guess there is no reason overlapping drains need to specify disjoint sets of machines -- either way, you should say something about that.

davidopp commented on docs/proposals/maintenance.md:
```diff
@@ -0,0 +1,216 @@
+multiple drains in flight at once.  If the drains have differing parameters,
+then more aggressive drains would generally starve safer ones.  While this could
+be a useful feature in emergencies, it will be a lot easier to reason about
+multiple drains in progress if they all share the same parameters, which would
```
Give an example of why it would be useful to have multiple drains if they all share the same parameters? (Seems you could collapse them into a single drain?)

davidopp commented on docs/proposals/maintenance.md:
````diff
@@ -0,0 +1,216 @@
+
+```
+type DrainSpec struct {
+  Nodes        NodeList
````
Add to "future" section of the doc that you could allow NodeList or a label selector here? For the 'one-shot drain' version it's reasonable to expect someone would do a kubectl command to fetch the list of nodes matching the desired selector and then pipe that to 'kubectl drain' (actually, you should probably mention that in that section) as you mentioned when we discussed this last week, but that doesn't seem as good for the server-driven approach, where I think people would eventually want a label selector to pick the nodes.

smarterclayton replied:
We typically would node this as an array of object references or a node selector where only one is specified at a time.

smarterclayton replied:
Model

mml replied:
What's the advantage of specifying this as a selector instead of a static list?  As an admin, it'd make me nervous that I might think I'm draining 8 machines but then something else happens and now I'm draining 800.

davidopp commented on docs/proposals/maintenance.md:
````diff
@@ -0,0 +1,216 @@
+}
+```
+
+A drain object would work much like a deployment.  A drain controller would
````
Deployment has a lot of other properties, like it creates other objects (like ReplicationControllers). This just seems to describe all controllers -- watch for the object type they manage, and work towards the desired end state. Was there some specific analogy to Deployment you were making, that isn't a general property of controllers?

smarterclayton replied:
A drain is like a cluster Job, that may also apply a strategy to the drain like a deployment.

mml replied:
> Was there some specific analogy to Deployment you were making, that isn't a general property of controllers?

Yes.  The property I care about is that it represents the management of a state transition over time:  there's a before and after state, and some rules about how to take the steps in between.  A deployment manages a transition from one replication controller to another, while a drain manages node state and the deletion of pods.

I'll clarify.

smarterclayton commented on docs/proposals/maintenance.md:
```diff
@@ -0,0 +1,216 @@
+multiple drains in progress if they all share the same parameters, which would
+be configured via kube-env and would represent the cluster admin's SLO.
+
+The drain object would thus look something like this
```
The modeling of this is the first example of an action being encoded as a rest resource, so we should be careful to ensure its right.  Another modeling might be as a subresource or field on node, where a client indicates a drain is requested and a controller executes a drain and then clears the field.  I'm not set on that, but this is a broad design implication that needs a deeper dive.

davidopp replied:
Why do you say "The modeling of this is the first example of an action being encoded as a rest resource"? If I create a ReplicationController, say with replicas=1, isn't that an action "create one pod" encoded as a rest resource?

The approach you suggested seems a little less declarative than what's suggested here. IIUC the approach you suggested would still need a rest resource to store the drain parameters; but instead of specifying a nodelist in that same resource, you'd request drains individually on each node you wanted to drain. Why do you think that is better? One problem is that I don't think specifying the nodes to drain via label selector would work in that approach. Also you'd have a lot of object writes (one to the subresource or field on each node).

smarterclayton replied:
I wasn't invested in the alternative, but want to have the design call out why it is a resource so we can extract that into API conventions and use it for future designs.  E.g this is a Job on the Cluster, it runs to completion but must record arguments state, has a termination state, is garbage collected by another controller eventually, etc.

thockin replied:
Is it really useful to have multiple in-flight drains or is it just a side-effect?  What if we modelled this as a (singleton?) Cluster object in the API and let that carry state or even point to ConfigData objects or something?

davidopp replied:
@thockin 

> Is it really useful to have multiple in-flight drains or is it just a side-effect? What if we modelled this as a (singleton?)

I think it might be reasonable to have multiple in-flight drains to disjoint sets of machines (for example, different dedicated machine groups might have different policies, which might be different from the policy for the shared pool--though maybe if you moved the drain spec into the collection, as suggested in #12611, then you wouldn't need this flexibility). 

BTW #7459 is relevant to your comment but I don't think it's super-closely-related.

thockin replied:
Yeah, I can buy that multiple drains makes sense in some cases.

On Sun, Nov 22, 2015 at 3:08 PM, David Oppenheimer <notifications@github.com
> wrote:

> In docs/proposals/maintenance.md
> <https://github.com/kubernetes/kubernetes/pull/17393#discussion_r45566418>
> :
>
> > +restarted with the same arguments, the end state would be exactly the same and
> > +we'd never break the shard strength rule, but the evictions might be more
> > +closely spaced.
> > +
> > +
> > +### Implementation of the above as a controller
> > +
> > +The parameters of a specific lame, drain, or unlame request could be pushed to
> > +the API server as a drain object.  The drain controller would have to manage
> > +multiple drains in flight at once.  If the drains have differing parameters,
> > +then more aggressive drains would generally starve safer ones.  While this could
> > +be a useful feature in emergencies, it will be a lot easier to reason about
> > +multiple drains in progress if they all share the same parameters, which would
> > +be configured via kube-env and would represent the cluster admin's SLO.
> > +
> > +The drain object would thus look something like this
>
> @thockin <https://github.com/thockin>
>
> Is it really useful to have multiple in-flight drains or is it just a
> side-effect? What if we modelled this as a (singleton?)
>
> I think it might be reasonable to have multiple in-flight drains to
> disjoint sets of machines (for example, different dedicated machine groups
> might have different policies, which might be different from the policy for
> the shared pool--though maybe if you moved the drain spec into the
> collection, as suggested in #12611
> <https://github.com/kubernetes/kubernetes/issues/12611>, then you
> wouldn't need this flexibility).
>
> BTW #7459 <https://github.com/kubernetes/kubernetes/issues/7459> is
> relevant to your comment but I don't think it's super-closely-related.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/17393/files#r45566418>.
>

mml replied:
Another reason for multiple drains is just that the need for them arises at different times and for different reasons.  For example, 10 machines are on faulty electrical equipment and we have to evacuate them in 30 minutes so the electrician can fix that.  In the background, we've been draining 30 other machines for non-urgent kernel upgrades.  I'll add an example like this.

This example conveniently avoids overlapping machines, which I will have to say something about.

smarterclayton commented on docs/proposals/maintenance.md:
```diff
@@ -0,0 +1,216 @@
+A drain object would work much like a deployment.  A drain controller would
+watch for drain objects and, within constraints, work toward the desired end
+state.  The action of the drain controller can be paused or stopped by deleting
+the drain object, and restarted by re-posting it.
```
How do multiple drain objects get handled?  What if they overlap with a drain in progress?

erictune commented on docs/proposals/maintenance.md:
```diff
@@ -0,0 +1,216 @@
+
+### Drain one node
+
+This is implemented in https://github.com/kubernetes/kubernetes/pull/16698
```
This implemented drain only with grace period.  

The next step after that could be "drain one node with shard-strength check".   That would help in two cases:
 
- you have some pet pods that you don't want to terminate.  you also don't want to remember which those are. Let  --min-shard-strength remember for you.
- a collection of pods temporarily is below desired replication count due to whatever reason. Don't make it worse.

erictune replied:
You could also include "time between evictions (of pods in the same replica set)" in that.  There could be several pods of the replica set on a node.  Spreading tries to prevent that, but it can certainly happen in some circumstances.

mml replied:
Since we have to implement the safety constraints anyway, is the implementation for one node any simpler than multi-node?

erictune commented on docs/proposals/maintenance.md:
```diff
@@ -0,0 +1,216 @@
+user specifying their requirements.  Around evictions, there are typically three
+parameters:
+
+- time between evictions (of pods in the same replica set)
```
It would be good to explain the reason for each of these service quality parameters.
e.g.: ensuring time between evictions allows a replacement pod some time to warm up and rebuild its caches before proceeding to the next victim; eviction notice allows a pod to write out a checkpoint, finish outstanding requests and background processing, etc;

mml replied:
> explain the reason for each of these service quality parameters

Done.

derekwaynecarr commented:
I may have missed this in another issue/discussion, but is there a milestone order for this?  For example, is the controller targeted for 1.2 or just the CLI command?

davidopp commented:
This actually isn't on any of our release milestones, partly because @mml is unsure how long he'll be working on Kubernetes. Clearly we'll be able to get #16698 into 1.2, but anything beyond that is still TBD. I think it's valuable to get the proposals written up either way, so he and/or others can work on the next steps once we figure out the priority.

pmorie commented:
Moving discussion to PR scope from: https://github.com/kubernetes/kubernetes/pull/17393#discussion_r45306233, since github does the incredibly annoying 'fold-outdated-diff' thing that makes this hard to find in the UI.

Let's get quorum on the final names for the verb to put yellow tape and slippery-when-wet signs around a node while under maintainence.  Quoting @mml:

> I would like a term that cleanly has two verbs and a noun describing the state. Do you have any suggestions along those lines? In the current case: lame, unlame, and lamed.
> Some suggestions from discussions with @bgrant0607 and @erictune 
> - cordon, uncordon, cordoned
> - decommission, recommission, decommissioned
> - close, open, closed (as in a faucet or business)
> - shut, reopen, shut (ditto)

My personal preference is for (1), cordon/uncordon/cordoned.

@mml @davidopp @erictune @bgrant0607 @derekwaynecarr thoughts?

mml commented:
> Let's get quorum on the final names for the verb

I like cordon/uncordon, and it's had the least (0?) objections.  If I don't hear anything by tomorrow morning, I'll rewrite this and #16698 with those names.

I don't think anyone objects to drain, but if there are objections (speak now or forever etc), evacuate is my second choice.

pmorie commented:
@mml I have no objections to drain

smarterclayton commented:
I'm fine with drain, cordon, and uncordon

On Thu, Dec 3, 2015 at 3:40 PM, Kubernetes Bot <notifications@github.com>
wrote:

> GCE e2e build/test *failed* for commit c39e4c2
> <https://github.com/kubernetes/kubernetes/commit/c39e4c2352da3e0aac262daee504cc09b83f6a8e>
> .
>
>    - Build Log
>    <https://storage.cloud.google.com/kubernetes-jenkins/pr-logs/c39e4c2352da3e0aac262daee504cc09b83f6a8e/kubernetes-pull-build-test-e2e-gce/18979/build-log.txt>
>    - Test Artifacts
>    <https://console.developers.google.com/storage/browser/kubernetes-jenkins/pr-logs/c39e4c2352da3e0aac262daee504cc09b83f6a8e/kubernetes-pull-build-test-e2e-gce/18979/artifacts/>
>    - Internal Jenkins Results
>    <http://goto.google.com/prkubekins/job/kubernetes-pull-build-test-e2e-gce//18979>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/17393#issuecomment-161777667>
> .
>

davidopp commented:
@mml is this ready to re-review?

mml commented:
No, I still have some comments to address, and it really needs a new revision to give more than lip service to some of these issues.

We do not have a strong story for everything that has been raised (e.g. how would complicated workflows work), and I'm not sure we really *need* that unless we plan to implement this soon, so I wonder where I should stop.  The unaddressed issues become fractal without a clear stopping point:  The more I write the more new things I uncover that aren't addressed.

bgrant0607 commented:
I'd like to wrap this up. Reassigning to myself to take a look, since I'm also reviewing #18263.

bgrant0607 commented:
Ah, this is mostly about SLOs. Ok, transferring back to @davidopp for now.

davidopp commented:
Yeah, there is no design or doc yet about using taints/tolerations in machine lifecycle events. I intentionally left that out of #18263 because I felt the dedicated node use case was sufficient justification for it and the amount of time to get a design doc approved seems to scale with the number of words in it. :-) @mml can write a separate doc about that once he has had time to think about it (and we get the taints/tolerations design doc merged).
